enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+metadata:
+  name: workers-us-east-1a
+  namespace: openshift-machine-api
+spec:
+  selector:
+    role: worker
+    zone: us-east-1a
+  remediationStrategy: delete
+  unhealthyConditions:
+    - name: Ready
+      status: false
+      timeout: 10m
+  maxUnavailable: 5
+status:
+  ExpectedHealthy: 10 // Total number of machines targeted by the label selector
+  CurrentHealthy: 10
+```
+
+TODO: how to define OR relation between unhealthyConditions, e.g `ready=false || ready=unknown`
+
+#### MachineHealthCheck controller:
+- Watch MachineHealthCheck resources and nodes.
+- Find machines backing nodes that meet the unhealthy criteria and signal remediaton systems with the chosen remediation strategy, e.g `healthchecking.openshift.io/remediation: reboot`.
+- If the remediation is "delete" then it runs the remediation logic and set the machine for deletion. 
+
+#### Out of tree remediation controller, e.g baremetal reboot:
+- Watch machines signaled for remediation with `healthchecking.openshift.io/remediation: reboot` and react as they need.
+
+### Risks and Mitigations
+
+
+## Design Details
+
+### Test Plan
+This feature will be tested for public clouds in the e2e machine API suite as the other machine API ecosystem https://github.com/openshift/cluster-api-actuator-pkg/tree/8250b456dec7b2fb06c591738518de1265e84a2c/pkg/e2e
+
+### Graduation Criteria
+An implementation of this feature is currently gated behind the `TechPreviewNoUpgrade` flag. This proposal wants to remove the gating flag and promote machine health check to a GA status with a beta API.
+
+### Upgrade / Downgrade Strategy
+
+The machine health check controller lives in the machine-api-operator image so the upgrades will be driven by the CVO which will fetch the right image version as usual. See:
+
+https://github.com/openshift/machine-api-operator/blob/474e14e4965a8c5e6788417c851ccc7fad1acb3a/install/0000_30_machine-api-operator_01_images.configmap.yaml
+
+https://github.com/openshift/machine-api-operator/blob/474e14e4965a8c5e6788417c851ccc7fad1acb3a/pkg/operator/operator.go#L222
+
+For supported out of tree remediation controllers, e.g baremetal reboot, the binary needs to be deployed in one of the available bare metal images or specify its own one. See:
+
+https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/baremetal_pod.go
+
+https://github.com/openshift/machine-api-operator/blob/474e14e4965a8c5e6788417c851ccc7fad1acb3a/pkg/operator/baremetal_pod.go
+
+### Version Skew Strategy
+
+## Implementation History
+
+Initial implementation:
+
+https://docs.google.com/document/d/1-RPiXfc33SyM7Gn-dogCWNigEWpSxXV_OZ7UKsQtI4E/edit#
+
+Related discussions:
+
+https://docs.google.com/document/d/10kauaJiXaWpvmd_qVsgIoZBNQLJMXMsQga3exV04ZTY/edit?ts=5d0b91e6#heading=h.4ifefbk4b4y2
+
+https://gist.github.com/bison/403bb921e1d5ed72f7edec2ccb47471c#remediationstrategy
+
+## Drawbacks
+
+## Alternatives
+
+Decoupling the administrator-defined short-circuiting logic.
```

Comment:
Does doing the base path proposal prevent this from maybe being possible in the future?  Not saying we would, but I don’t see health check being a blocker (except due to implementation).

Agree with pros and cons

Reply:
mm it does not prevent it necessarily. The `maxUnavailable` field proposed here https://github.com/openshift/enhancements/pull/9/files#diff-53036ce0d8589cdbed15ed2e95076b54R105 would need to be deprecated in favour of the semantics of the additional CRD

Reply:
worth to say that we already have this de-coupling